Explore: Add documentation around transformations in the explore editor (#78082)
grafana/grafana: docs/sources/explore/correlations-editor-in-explore.md

Reviewer: I think the rename on this page can be avoided because as a task topic, we would like this page to have a title containing a verb and the object.

Previously "Use" was the verb and "correlations editor in Explore" was the object.

I believe this is a "task topic" and should conform to the advice in https://grafana.com/docs/writers-toolkit/structure/topic-types/task/.

Reviewer: Alternatively, 

Perhaps the task based instructions should remain in the original file and the concept information here should go in this file?

It would also have the benefit of making the changes you have made easier to understand.

Author: My thought was that concept information would be under administration where it was originally. Explore would be only about the explore editor and refer back to the concept information as a reference. I don't understand what you mean by "It would also have the benefit of making the changes you have made easier to understand."

Author: > I think the rename on this page can be avoided because as a task topic, we would like this page to have a title containing a verb and the object.

The reason I feel like we should both move and change this page is because with the addition of the transformation section, this feature becomes a lot less straightforward. Prior to this feature, it was a fairly simple builder where one could point and click to create a correlation (thus it being a task topic on its own). With the addition of transformations, there is an optional, more complex sub topic to creating correlations that I think deserves a separate section to make clear that is may not be needed. A small, straightforward thing for explore that lives in administration makes sense. Something more complicated for explore should be in explore and crosslinked.

So for this page, I'd like to keep it not a task topic on its own, but rather have a couple topic tasks on the same page. It doesnt seem like quite enough content for multiple pages.  I'll clean up the task topics I have here to match the advice.

Reviewer 2: consolidating multiple tasks into a single page is fine so long as the tasks are related and the title is general enough to encompass both.

I agree with both of you to keep the task titles following the standard verb + object format.

## Create a correlation

1. Go to Explore page.
1. Select a data source that you would like to be [the source data source]({{< relref "../administration/correlations/correlation-configuration#source-data-source-and-result-field" >}}) for a new correlation.
1. Run a query producing data in [a supported visualization]({{< relref "../administration/correlations#correlations" >}}).
1. Click "+ Add" button in the top toolbar and select "Add correlation" (you can also select "Correlations Editor" from the [Command Palette]({{< relref "../search#command-palette" >}})).
1. Explore is now in Correlations Editor mode indicated by a blue border and top bar. You can exit Correlations Editor using the Exit button in the top bar.
1. The visualization is enriched with links allowing to create new correlations. Links are attached to the data that you can use to build a new query:
   - Logs: links are displayed next to field values inside log details for each log row
   - Table: every table cell is a link

> 1. Click on a link to begin adding new correlation. Links are associated with a field that will be used as [a results field of a correlation]({{< relref "../administration/correlations/correlation-configuration" >}}). This means the correlation link will be created on the link you select once the correlation is saved.

Reviewer: nit: prefer "after" over "once": https://developers.google.com/style/word-list#once.

```suggestion
1. Click on a link to begin adding new correlation. Links are associated with a field that will be used as [a results field of a correlation]({{< relref "../administration/correlations/correlation-configuration" >}}). The correlation link will be created on the link you select after the correlation is saved.
```

1. Explore opens a split view. Use the right pane to setup [the target query source of the correlation]({{< relref "../administration/correlations/correlation-configuration#target-query" >}}).
1. Build target query using [variables syntax]({{< relref "../dashboards/variables/variable-syntax" >}}) with variables from the list provided at the top of the pane. The list contains sample values from the selected data row.

> 1. Provide optional label and description. Label is used as the name of the link inside the visualization. It can contain variables.

Reviewer 3: I think we could add an optional step to indicate transformations can be added at this stage: "Provide optional transformations (see below for more details)."

1. Click "Save" button in the top bar to save the correlation and exit Correlations Editor mode.

> 1. The link used to create the correlation will be replaced with a data link in each row. The variables will be dynamically replaced with values from a selected row.

Reviewer: Avoid "will" unless talking about future behavior: https://developers.google.com/style/word-list#will

I don't think I understand the second sentence: The variables are dynamically replaced with values from a selected row.

```suggestion
1. The link used to create the correlation is replaced with a data link in each row. The variables are dynamically replaced with values from a selected row.
```

Author: > I don't think I understand the second sentence: The variables are dynamically replaced with values from a selected row.

This is the most complicated idea about correlations to get across and I'm struggling how to construct a concise sentence describing it without an example.

Basically you end up with data with a button alongside it. 
Like `field1=hello field2=world`

Building the correlation, you construct a query with a variable
`SELECT * FROM ${field1} WHERE name=${field2}`

When you click the button, data next to the button goes into the constructed query. 
`SELECT * FROM hello WHERE name=world`

Maybe something like

```The link used to create the correlation is replaced with a data link in each row. When the link is clicked, the query defined will run in another pane, with the variables replaced dynamically with the values from the selected row.```

Reviewer: I like that explanation a lot. I can see that it's a complicated thing to explain written down but at least from my perspective as an engineer, that rewrite makes sense to me!

## Transformations

> Transformations allow us to extract values that exist in a field with other data. A prime example is getting one portion of a log line out to use in a correlation. For more details on transformations in correlations, see the [correlations documentation]({{< relref "../administration/correlations/correlation-configuration/#correlation-transformations" >}}).

Reviewer 3: ```suggestion
Transformations allow you to extract values that exist in a field with other data. A prime example is getting one portion of a log line out to use in a correlation. For more details on transformations in correlations, see the [correlations documentation]({{< relref "../administration/correlations/correlation-configuration/#correlation-transformations" >}}).
```

https://grafana.com/docs/writers-toolkit/write/style-guide/style-conventions/#address-users-clearly

> After clicking one of the generated links in the editor mode, you can add transformations by clicking the "Add transformation" button in the "Transformations" dropdown. This will pop up a modal, and you can use a transformation in your correlation with the following steps:

Reviewer: Attempting to reword to avoid "will".

Also, I believe we prefer dialog box over modal.

```suggestion
After clicking one of the generated links in the editor mode, you can add transformations by clicking the "Add transformation" button in the "Transformations" dropdown.
A dialog box appears, and you can use a transformation in your correlation with the following steps:
```

> 1. First, select a field that the transformation will be applied. This should be a field where we only want to use sections of the field, such as a log line. Once selected, the value of that field will show up to assist building the transformation.

Reviewer: Avoiding "will" and using the second person.

```suggestion
1. First, select a field to apply the transformation to.
  This should be a field where you only want to use sections of the field, such as a log line.
  Once selected, the value of that field shows up to assist building the transformation.
```

1. Select the type of the transformation. See the [correlations documentation]({{< relref "../administration/correlations/correlation-configuration/#correlation-transformations" >}}) for the options and relevant settings.

> 1. Based on what's selected, you may immediately see one or more variables populate, or you may need to fill in more fields

Reviewer 3: ```suggestion
1. Based on what's selected, you may immediately see one or more variables populate, or you may need to fill in more options
```

There's a bit of ambiguity "form field" vs "data frame field" (step 1). Maybe "options" would be more clear here?

1. Select `Add transformation to correlation` to add the specified variables to the list of variables in the list of variables available.

### Notes for regular expressions

> For regular expressions in this modal, the `mapValue` referred to in other documentation is called `Variable Name` here. The field value will highlight with the matching text based on the expression. Use regular expression capture groups to select what portion of the match should be extracted. When a valid regular expression is provided, you will see the variable and the value of that variable appear below the `Variable Name` field.

Reviewer: I'm not sure I understand "For regular expressions in this modal, the `mapValue` referred to in other documentation is called `Variable Name` here"
```suggestion
For regular expressions in this dialog box, the `mapValue` referred to in other documentation is called `Variable Name` here.
The field value highlights the matching text based on the expression.
Use regular expression capture groups to select what portion of the match should be extracted.
When a valid regular expression is provided, the variable and the value of that variable appear below the `Variable Name` field.
```

Author: > I'm not sure I understand "For regular expressions in this modal, the mapValue referred to in other documentation is called Variable Name here"

MapValue is the original term we used in the provisioning and original correlations editor. With this editor being more user friendly in what it was doing, we decided to use a more user friendly name. See https://github.com/grafana/grafana/blob/main/docs/sources/administration/correlations/correlation-configuration/index.md?plain=1#L99 for the use of mapValue in other documentation

Author: Changing
>  The field value will highlight with the matching text based on the expression

to 

> The field value highlights the matching text based on the expression.

Doesn't quite make sense in my opinion. As you type an expression, a highlight shows according to the match on the expression. The field value is highlighted, it doesn't do the highlighting.

https://github.com/grafana/grafana/assets/1533128/03199793-dd23-4b71-b9a5-a2cc1b412a62

Maybe "Any text that matches the expression in the field value is highlighted." ?

Reviewer: I see what you are saying but I don't think my suggestion changes the existing meaning.
The original text "The field value will highlight" also suggests that the field value is doing the highlighting.

I think your rewrite is more technically accurate but is in passive voice. In general we try and prefer active voice but it is harder to rewrite this actively because I'm not actually sure what thing does the highlighting. Perhaps:

> Grafana highlights any text that matches the expression in the field value.

## Example: Create a text to graph correlation

This example shows how to create a correlation using Correlations Editor in Explore.

Correlations allow you to use results of one query to run a new query in any data source. In this example you will run a query that renders tabular data. The data will be used to run a different query that yields a graph result.

Please make sure you have set up [a test data source]({{< relref "../datasources/testdata#testdata-data-source" >}}).

1. Go to Explore.
1. Select `test data source`.
1. Click on "+ Add" dropdown and select "Add correlation" button.
1. Explore is now in Correlations Editor mode indicated by a blue border.
1. Select scenario: `CSV File`.
1. Select file: `population_by_state.csv`.
1. Note that each cell is a link that you can click on to start creating a new correlation.

   {{< figure src="/static/img/docs/correlations/screenshot-correlations-editor-source-10.2.png" max-width="600px" caption="Selecting the source of a correlation" >}}

> 1. Create new correlation that attaches a data link in the `State` column: click on any cell in the `State` column, e.g. "California".

Reviewer: https://developers.google.com/style/word-list#eg

```suggestion
1. Create new correlation that attaches a data link in the `State` column: click on any cell in the `State` column, for example, "California".
```

1. Explore opens in split view. Select the same data source you selected in the left pane.
1. The helper above the query editor contains all available variables you can use the target query. Variables contain all data fields (table columns) from the selected row.
1. Select Scenario: `CSV Metric Values`.
1. In the Query Editor's `String Input` field provide variables with population values for each year: `${1980},${2000},${2020}`. This will generate a graph using variable values.
1. In the Query Editor's `Alias` field, write `${State}`

   {{< figure src="/static/img/docs/correlations/screenshot-correlations-editor-target-10.2.png" max-width="600px" caption="Setting up the target of a correlation" >}}

1. Run a query to see that it produces a graph using sample values from the variables.
1. Save the correlation using "Save" button in the top.

> Once the correlation is saved Explore will exit Correlations Editor automatically and re-rerun the query in the left pane. By clicking a state name, the query on the right is re-ran with values from the row being inserted into the CSV, thus changing the graph. The query is rerun with updates values every time you click on a state name.

Reviewer: ```suggestion
After the correlation is saved Explore exits Correlations Editor automatically and rerun the query in the left pane.
By clicking a state name, the query on the right is rerun with values from the row being inserted into the CSV, thus changing the graph.
The query is rerun with updates values every time you click on a state name.
```

{{< figure src="/static/img/docs/correlations/screenshot-correlations-example-link-10.2.png" max-width="600px" caption="Result of clicking on a data link" >}}

You can apply the same steps to any data source. Correlations allow you to create links in visualizations to run dynamic queries based on selected data. In this example we used data returned by a query to build a new query generating different visualization using the same data source. However, you can create correlations between any data sources to create custom exploration flows.

## Example: Create a logs to table correlation

> For this example, we will create a correlation to demonstrate how to use transformations to extract values from the log line and another field.

Reviewer: ```suggestion
In this example, you will create a correlation to demonstrate how to use transformations to extract values from the log line and another field.
```

1. Go to Explore.
1. Select `test data source`.
1. Click on "+ Add" dropdown and select `Add correlation` button.
1. Explore is now in Correlations Editor mode indicated by a blue border.
1. Select scenario: `Logs`.
1. Expand a log line to see the correlation links. Select `Correlate with hostname`.
1. Explore opens in split view. Select the same data source you selected in the left pane.
1. The helper above the query editor contains all available variables you can use the target query.
1. Expand the transformations section, and click the `Add transformation` button.
1. Under the `Field` dropdown, select `message`. See the log line shows up as example data.
1. Under `Type`, select `Logfmt`. See the list of variables populate.
1. Click `Add transformation to correlation`.
1. Again click the `Add transformation` button and under `Field`, select `hostname`.
1. Select `Regular expression` under `Type`.
1. Under `Expression`, type `-([0-9]*)`. This selects any numbers to the right of the dash.
1. Under `Variable Name`, type `hostNumber`
1. See the variable populate in the variable list.
1. Click `Add transformation to correlation` to add it to the other variables.

> 1. In the datasource editor, select Scenario: `CSV Content`.

Reviewer: Unless the UI says "datasource"

```suggestion
1. In the data source editor, select Scenario: `CSV Content`.
```

1. In the text box below, provide the following and save the correlation.

```
time,msg,hostNumber,status
${time},${msg},${hostNumber},${status}
```

> This will close the split view and re-run the left query. Expand any log line to see the correlation button. Clicking it will pull up the split view with the `time` (a field), `msg` (extracted with `logfmt` from the log line), host number (extracted with `regex` from the `hostname`) and the `status` (extracted with `logfmt` from the log line).

Reviewer: ```suggestion
This closes the split view and reruns the left query.
Expand any log line to see the correlation button.
Clicking the correlation button opens the split view with the `time` (a field), `msg` (extracted with `logfmt` from the log line), host number (extracted with `regex` from the `hostname`) and the `status` (extracted with `logfmt` from the log line).
```
